### PR TITLE
allow a dict in DataFrame.astype()

### LINKED ISSUE #411: error: Argument 1 to "astype" of "DataFrame" has incompatible type "Dict[str, str]" [arg-type]
```python
import pandas as pd

df = pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})
columns_types = {"a" : "int", "b" : "float"}
df = df.astype(columns_types)  # ✘ raises [arg-type]
```

Context:

```
error: Argument 1 to "astype" of "DataFrame" has incompatible type "Dict[str, str]"; expected
"Union[str, Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]], Dict[str, Union[str, Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]]]]]"  [arg-type]
    df = df.astype(columns_types)  # ✘ raises [arg-type]
                   ^~~~~~~~~~~~~
note: "Dict" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
note: Consider using "Mapping" instead, which is covariant in the value type
```

Suggestion: 

```python
    def astype(
        self,
        dtype: _str | Dtype | dict[_str, _str | Dtype],
        copy: _bool = ...,
        errors: _str = ...,
    ) -> DataFrame: ...
```

needs to be replaced with

```python
    def astype(
        self,
        dtype: _str | Dtype | Mapping[_str, _str | Dtype],
        copy: _bool = ...,
        errors: _str = ...,
    ) -> DataFrame: ...
```

**Please complete the following information:**
- python 3.10.6
- mypy                          0.982
- pandas                        1.5.1
- pandas-stubs                  1.5.1.221024

### PR DESCRIPTION
**Pull request by Dr-Irv**
- [x] Closes #411
- [x] Tests added: 
  - `test_frame.py:test_astype_dict()`

### CODE REVIEW

File: pandas-stubs/core/frame.pyi [hunk 1]
```diff
@@ -1357,7 +1357,7 @@ class DataFrame(NDFrame, OpsMixin):
     ) -> DataFrame: ...
     def astype(
         self,
-        dtype: _str | Dtype | dict[_str, _str | Dtype] | Series,
+        dtype: _str | Dtype | Mapping[_str, _str | Dtype] | Series,
```
Comment (twoertwein) [inline pandas-stubs/core/frame.pyi:1360]: the keys can probably be any `Hashable(T)` as the columns might not be `str`.

Comment (Dr-Irv) [reply pandas-stubs/core/frame.pyi]: done in https://github.com/pandas-dev/pandas-stubs/pull/449/commits/5ff40dedffb393b03277c03492df9001e7fb1b23

**Review by twoertwein — APPROVED** [review]

File: pandas-stubs/core/frame.pyi (hunk 1 above)
Comment (twoertwein) [reply pandas-stubs/core/frame.pyi]: Could maybe add a test for a non-str column

**Comment by twoertwein** [comment]
Thank you @Dr-Irv !

**Pull request merged**